### PR TITLE
[SPARK-5123][SQL] Reconcile Java/Scala API for data types.

### PR DESCRIPTION
**Pull request by rxin**
Having two versions of the data type APIs (one for Java, one for Scala) requires downstream libraries to also have two versions of the APIs if the library wants to support both Java and Scala. I took a look at the Scala version of the data type APIs - it can actually work out pretty well for Java out of the box.

As part of the PR, I created a sql.types package and moved all type definitions there. I then removed the Java specific data type API along with a lot of the conversion code.

This subsumes https://github.com/apache/spark/pull/3925

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/dataTypes.scala [hunk 1]
```diff
@@ -237,10 +256,12 @@ object NativeType {
     StringType -> 4096)
 }
 
+
 trait PrimitiveType extends DataType {
   override def isPrimitive = true
 }
 
+
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/dataTypes.scala]: Double newlines here?

File: sql/core/src/main/scala/org/apache/spark/sql/api/java/UDFRegistration.scala [hunk 2]
```diff
@@ -94,159 +93,159 @@ private[java] trait UDFRegistration {
...
 
   // scalastyle:off
-  def registerFunction(name: String, f: UDF1[_, _], dataType: DataType) = {
-    val scalaType = asScalaDataType(dataType)
+  def registerFunction(
+      name: String, f: UDF1[_, _], @transient dataType: DataType) = {
     sqlContext.functionRegistry.registerFunction(
       name,
-      (e: Seq[Expression]) => ScalaUdf(f.asInstanceOf[UDF1[Any, Any]].call(_: Any), scalaType, e))
+      (e: Seq[Expression]) => ScalaUdf(f.asInstanceOf[UDF1[Any, Any]].call(_: Any), dataType, e))
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/api/java/UDFRegistration.scala]: Not in this PR, but we should add `TypeTags` to these functions for the input arguments.

**Comment by marmbrus** [comment]
Okay, the other PR is in, if you fix the merge conflicts this LGTM.  We also need to figure out how to publish scala/java docs here.

**Comment by rxin** [comment]
I just rebased. Let's see ...

**Comment by rxin** [comment]
Jenkins, retest this please.

**Comment by JoshRosen** [comment]
You might want to copy the commit message over from the other PR.

**Comment by rxin** [comment]
done - thanks for reminding, josh.

**Comment by rxin** [comment]
Ok merging in master. Thanks for looking at this.

**Pull request closed without merging**